### ISSUE TITLE
Set local address to bind a specific interface

### ISSUE THREAD

**Issue by amunra** [algesten/ureq#692]
I'm more or less looking for the equivalent of https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html#method.local_address

This is useful for cases where there are multiple NICs installed and one of them is preferred (e.g., because it supports kernel bypass).

Looking at the code in `stream.rs` it looks like this is the main area of the code:

```rust
        let stream = if proto.is_some() && Some(Proto::HTTP) != proto {
            connect_socks(
                unit,
                proxy.clone().unwrap(),
                connect_deadline,
                sock_addr,
                hostname,
                port,
                proto.unwrap(),
            )
        } else if let Some(timeout) = timeout {
            TcpStream::connect_timeout(&sock_addr, timeout)
        } else {
            TcpStream::connect(sock_addr)
        };
```

If I understand it correctly, the limiting factor here is that `std::net::tcp::TcpStream` doesn't (as far as I can tell) support binding before connecting.

In other words, this would require [socket2](https://crates.io/crates/socket2).
Alternatively, I guess, it could be approached with a custom `Connector` trait.

Is this something that might make sense adding as a feature?

**Comment by algesten** [comment]
Hi @amunra!

Because we want to be lean, I don't see us relying on socket2. However a connector trait has been proposed in #518. One way forward would be to explore the issues in that PR.

**Comment by amunra** [comment]
Yeah, that makes sense. I'll think about this and see if I can come up with something that will cater for the multiple requirements in that PR's discussion.

I'll continue there and close this issue.
